// ng2-dragula/__ivy_ngcc__/components/dragular.module.js
"use strict";
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./dragula.directive');
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var core_1 = require("@angular/core");
var dragula_directive_1 = require("./dragula.directive");
var dragula_provider_1 = require("./dragula.provider");
var DragulaModule = (function () {
    function DragulaModule() {
    }
DragulaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DragulaModule });
DragulaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DragulaModule_Factory(t) { return new (t || DragulaModule)(); }, providers: [dragula_provider_1.DragulaService] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DragulaModule, { declarations: [ɵngcc1.DragulaDirective], exports: [ɵngcc1.DragulaDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DragulaModule, [{
        type: core_1.NgModule,
        args: [{
                exports: [dragula_directive_1.DragulaDirective],
                declarations: [dragula_directive_1.DragulaDirective],
                providers: [dragula_provider_1.DragulaService]
            }]
    }], function () { return []; }, null); })();
    return DragulaModule;
}());
exports.DragulaModule = DragulaModule;
