// ng-click-outside/__ivy_ngcc__/lib/click-outside.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var click_outside_directive_1 = require("./click-outside.directive");
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./click-outside.directive');
var ClickOutsideModule = /** @class */ (function () {
    function ClickOutsideModule() {
    }
    /** @nocollapse */
    ClickOutsideModule.ctorParameters = function () { return []; };
ClickOutsideModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClickOutsideModule });
ClickOutsideModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClickOutsideModule_Factory(t) { return new (t || ClickOutsideModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClickOutsideModule, { declarations: [ɵngcc1.ClickOutsideDirective], exports: [ɵngcc1.ClickOutsideDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClickOutsideModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [click_outside_directive_1.ClickOutsideDirective],
                exports: [click_outside_directive_1.ClickOutsideDirective]
            }]
    }], function () { return []; }, null); })();
    return ClickOutsideModule;
}());
exports.ClickOutsideModule = ClickOutsideModule;
